pkg/_pub_shared/lib/utils/http.dart

```diff
@@ -23,15 +23,82 @@ http.Client httpRetryClient({
   int? retries,
   bool lenient = false,
 }) {
-  return RetryClient(
+  return _PubRetryClient(
     innerClient ?? http.Client(),
-    when: (r) =>
-        (lenient && r.statusCode >= 500) ||
-        _transientStatusCodes.contains(r.statusCode),
-    retries: retries ?? 5,
-    // TODO: Consider implementing whenError to handle DNS + handshake errors.
-    //       These are safe, retrying after partially sending data is more
-    //       sketchy, but probably safe in our application.
-    whenError: (e, st) => lenient || e is SocketException,
+    retries ?? 1,
+    lenient,
+    innerClient != null,
   );
 }
+
+class _PubRetryClient extends http.BaseClient {
+  final http.Client _inner;
+  final int _attempts;
+  final bool _closeInner;
+  final bool _lenient;
+
+  _PubRetryClient(
+    this._inner,
+    this._attempts,
+    this._lenient,
+    this._closeInner,
+  );
+
+  @override
+  Future<http.StreamedResponse> send(http.BaseRequest request) async {
+    final requestBody = await request.finalize().toBytes();
+    for (var i = 0; i < _attempts; i++) {
+      try {
+        // wait between attempts
+        await Future.delayed(Duration(milliseconds: i * 100));
+
+        // send new request through inner client
+        final rs = await _inner.send(_copyRequest(request, requestBody));
+
+        // collect response until the entire body is recieved
+        final responseBody = await rs.stream.toBytes();
```

Comment:
Is it not a problem that we are no longer streaming?

Reply:
Yes, that is an important point, we need to review the endpoints that may take the http stream and pass it directly to the client calling `pub.dev`. We could still use retries in such cases, if the content is known to be below a certain threshold.